### PR TITLE
Add a fork method to Drop

### PR DESCRIPTION
**Pull request by chrisvest**
Motivation:
We have multiple places where ownership of some memory, and its associated drop instance, branches.
These were previously handled by carefully threading drops through ArcDrop instances.
We can avoid these careful gymnastics and make the code clearer by adding an explicit fork() method to the Drop interface.
This in turn allows the CleanerDrop to wrap the ArcDrop, which logically makes more sense.

Modification:
Add a fork method to the Drop interface.
The Drop interface is no longer a functional interface.
Create some static methods to standardise the way drops are created and wrapped, so these are always done in the same way.
The CleanerDrop is now the outer-most drop instance, which and each fork of the CleanerDrop creates a new CleanerDrop instance.
This allows us to theoretically track leaks more closely associated with the specific buffer or object that leaked, rather than marking the original allocation as leaked when any split-off or child-buffer is leaked.

Result:
Cleaner code, and more precise leak tracked (once we have leak tracking).

This PR is based on #11710

### CODE REVIEW

**Comment by hyperxpro** [comment]
Do you have full documentation for the new Buffer implementation?

**Comment by chrisvest** [comment]
@hyperxpro Only what's in the javadoc.

File: buffer/src/main/java/io/netty/buffer/api/pool/PoolThreadCache.java [hunk 1]
```diff
@@ -59,27 +59,28 @@
...
+            // Only check if there are caches in use.
+            if ((smallSubPageCaches != null || normalCaches != null)
+                && freeSweepAllocationThreshold < 1) {
+                throw new IllegalArgumentException("freeSweepAllocationThreshold: "
+                                                   + freeSweepAllocationThreshold + " (expected: > 0)");
+            }
+
+            cacheRef = new WeakReference<>(new Cache(arena, smallSubPageCaches, normalCaches));
+            arenaReferenceCounter = arena.numThreadCaches;
+            arenaReferenceCounter.getAndIncrement();
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/api/pool/PoolThreadCache.java:79]: Is this change related ? Seems like no.

Comment (chrisvest) [reply buffer/src/main/java/io/netty/buffer/api/pool/PoolThreadCache.java:79]: This is from #11710, which this PR is based on.

File: buffer/src/main/java/io/netty/buffer/api/bytebuffer/NioBuffer.java [hunk 2]
```diff
@@ -62,14 +62,15 @@
...
-        super(new ArcDrop<>(ArcDrop.acquire(parent.unsafeGetDrop())), parent.control);
+        super(parent.unsafeGetDrop().fork(), parent.control);
         base = parent.base;
         rmem = bbslice(parent.rmem, 0, parent.rmem.capacity()); // Need to slice to get independent byte orders.
         assert parent.wmem == CLOSED_BUFFER;
         wmem = CLOSED_BUFFER;
         roff = parent.roff;
         woff = parent.woff;
         constBuffer = true;
+        unsafeGetDrop().attach(this);
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/api/bytebuffer/NioBuffer.java:73]: calling `this` from within a constructor is kind of a bad-pattern as it may be possible to pass down a object which is not fully constructed yet. Consider using a static method that will first create the `NioBuffer` and then attach it.

File: buffer/src/main/java/io/netty/buffer/api/unsafe/UnsafeBuffer.java [hunk 3]
```diff
@@ -77,6 +81,7 @@
         roff = parent.roff;
         woff = parent.woff;
         constBuffer = true;
+        unsafeGetDrop().attach(this);
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/api/unsafe/UnsafeBuffer.java:84]: same comment as above.

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Comment by chrisvest** [comment]
@normanmaurer Comments addressed. Please take a look.

**Review by normanmaurer — APPROVED** [review]
